ldap/servers/slapd/log.c

```diff
@@ -3013,137 +3204,969 @@ log__open_accesslogfile(int logfile_state, int locked)
+    slapi_pblock_get(pb, SLAPI_OPERATION, &operation);
+    slapi_pblock_get(pb, SLAPI_BIND_TARGET_SDN, &bind_sdn);  // this is not correct for mod ops
+    slapi_pblock_get(pb, SLAPI_CONN_DN, &binddn);
+    slapi_pblock_get(pb, SLAPI_ORIGINAL_TARGET_DN, &target_dn);
+    slapi_pblock_get(pb, SLAPI_REQUESTOR_ISROOT, &isroot);
+    slapi_pblock_get(pb, SLAPI_BIND_METHOD, &method);
+    slapi_pblock_get(pb, SLAPI_BIND_SASLMECHANISM, &saslmech);
+
+    conn_id = pb_conn->c_connid,
+    op_id = operation->o_opid,
+    client_ip = pb_conn->c_ipaddr;
+    server_ip = pb_conn->c_serveripaddr;
+    ldap_version = pb_conn->c_ldapversion;
+    if (saslmech) {
+        external_bind = !strcasecmp(saslmech, LDAP_SASL_EXTERNAL);
+    }
+
+    if (strcmp(event_type, SECURITY_AUTHZ_ERROR) == 0) {
+        if (strlen(target_dn) > 500) {
+            PR_snprintf(authz_target, sizeof(authz_target), "target_dn=(%.500s...)", target_dn);
+        } else {
+            PR_snprintf(authz_target, sizeof(authz_target), "target_dn=(%s)", target_dn);
+        }
+        msg_text = authz_target;
+    } else {
+        /* For authz events we need the connection DN, but for all other events we use
+         * bind target */
+        binddn = (char *)slapi_sdn_get_dn(bind_sdn);
+    }
+    if (binddn == NULL) {
+        binddn = "";
+    }
+    /* Determine the bind method and mechanism */
+    switch (method) {
+    case LDAP_AUTH_SASL:
+        if (external_bind && pb_conn->c_client_cert) {
+            /* Client Certificate */
+            PR_snprintf(method_and_mech, sizeof(method_and_mech), "TLSCLIENTAUTH");
+        } else if (external_bind && pb_conn->c_unix_local) {
+            /* LDAPI */
+            PR_snprintf(method_and_mech, sizeof(method_and_mech), "LDAPI");
+        } else if (!strcasecmp(saslmech, "GSSAPI") || !strcasecmp(saslmech, "DIGEST-MD5")) {
+            /* SASL */
+            PR_snprintf(method_and_mech, sizeof(method_and_mech), "SASL/%s", saslmech);
+        }
+        break;
+    default:
+        /* Simple auth */
+        PR_snprintf(method_and_mech, sizeof(method_and_mech), "SIMPLE");
+    }
+
+    /* Get the time */
+    struct timespec tsnow;
+    if (clock_gettime(CLOCK_REALTIME, &tsnow) != 0) {
+        /* Make an error */
+        PR_snprintf(local_time, sizeof(local_time), "vslapd_log_security, Unable to determine system time");
+        log__error_emergency(local_time, 1, 0);
+        return -1;
+    }
+    if (format_localTime_hr_log(tsnow.tv_sec, tsnow.tv_nsec, sizeof(local_time), local_time, &ltlen) != 0) {
+        /* MSG may be truncated */
+        PR_snprintf(local_time, sizeof(local_time), "vslapd_log_security, Unable to format system time");
+        log__error_emergency(local_time, 1, 0);
+        return -1;
+    }
+
+
+    /* build the format string, and truncate really long DN's */
+    strcpy(fmtstr, "{\"date\": \"%s\", \"utc_time\": %ld.%.09ld, \"event\": \"%s\", \"dn\": ");
+    if (slapi_sdn_get_size(bind_sdn) > 512) {
```

Comment:
This opens us up to json injection and tampering attacks due to the fact that dn (bindsdn) is remotely controlled, meaning that this can cause log forgery. We can't use strcat/fmtstrs with remotely controlled inputs like this, we probably need to use a proper json library so write these strings so they are escaped correctly.

Reply:
I thought since all these values have come through openldap ber decoding they should be safe.  But using a real library for the JSON building would be nice.  Have to look into that.

Reply:
I am afraid that ber decoding function cannot help against injection because you are able to encode/decode any binary data through ber.  @Firstyear is right, the DN should be escaped.

Reply:
I wasn't disagreeing with @Firstyear, I really thought openldap validated DN's and would reject them if they were invalid. 
 Well then our other logs are subject to this because this is exactly how we print the DN in the access log (from the pblock target bind dn).  Or is this specific to JSON?

Reply:
Although libber does not perform any check about tthe DN (it is an octet string at this level), the string that is stored in the security record is a valid DN because slapi_sdn_get_dn does normalize the DN and returns NULL if the DN is not parsable.
But a valid DN may contain characters that requires escape like \ otherwise it will be lost when decoding the JSON value and the security application wanting to use this value dn will not be able to parse the dn.

BTW It should probably be a good idea to add a test user like cn=Test with \"quotes\" in bind dn,dc=example, dc=foo 
    in your pytest test then to bind on this dn and see if the dn within the security log could be parsed as a dn 
     and is the expected dn.

Reply:
@mreynolds389 openldap makes sure the DNs are valid for ldap, but that doesn't mean they can't contain characters that have meanings in json. :)

Reply:
Ok ok you guys can stop beating me up on this.  I thought openldap did all the necessary escaping.  I was wrong  :-)

Reply:
🔨

Reply:
It's all good mate, there are tons of layers involved so it's good we picked it up now and not later!